server/etcdserver/v3_server.go

```diff
@@ -279,6 +289,12 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
 }
 
 func (s *EtcdServer) LeaseRenew(ctx context.Context, id lease.LeaseID) (int64, error) {
+	if s.isLeader() {
+		if err := s.waitApplyPendingCommits(); err != nil {
```

Comment:
Noticed that this wait function doesn't handle context cancellation. Maybe we could hit two birds with one stone by setting timeout on context and using context.Done instead of sleep in `waitApplyPendingCommits` (might require some changes in error handling).

Reply:
context can work, but I don't think it is necessary here. Usually it's useful when there is a deep call stack or for controlling of separate goroutines. 

It can also complicate the error handling.  The current implementation is straightforward and simple enough.

Reply:
I think we should still add case for `ctx.Done()` in `waitApplyPendingCommits` select to handle client request cancel.

Reply:
Got your point now, you are talking about the context for **client** request . Yes, makes sense. Will update.

Reply:
Done.